packages/devtools_app/lib/src/shared/memory/retaining_path.dart

```diff
@@ -0,0 +1,216 @@
+  int stored = 0;
+
+  /// Only displayed paths are stringified.
+  int stringified = 0;
+}
+
+/// A retaining path from the root to an object.
+///
+/// This class is used to represent the shortest retaining path from the root to an object.
+///
+/// Equal paths are not stored twice in memory.
+/// The path does not include the retained object itself.
+///
+/// The retaining path is represented as a list of classes, ignoring information
+/// about concrete instances and fields.
+/// To get more detailed information about the retaining path for a specific object,
+/// use [`leak_tracker/formattedRetainingPath`](https://github.com/dart-lang/leak_tracker/blob/f5620600a5ce1c44f65ddaa02001e200b096e14c/pkgs/leak_tracker/lib/src/leak_tracking/helpers.dart#L58).
+class PathFromRoot {
+  PathFromRoot._(this.path)
+      : assert(() {
+          usage.constructed++;
+          return true;
+        }()),
+        hashCode = path.isEmpty ? _hashOfEmptyPath : Object.hashAll(path),
+        classes = calculateSetOfClasses ? path.toSet() : const {};
+
+  /// For objects directly referenced from root.
+  const PathFromRoot._empty()
+      : path = const [],
+        classes = const {},
+        hashCode = _hashOfEmptyPath;
+
+  factory PathFromRoot.forObject(
+    HeapSnapshotGraph graph, {
+    required List<int> shortestRetainers,
+    required int objectId,
+    required int heapRootIndex,
+  }) {
+    var nextObjectId = shortestRetainers[objectId];
+    if (nextObjectId == heapRootIndex) {
+      return empty;
+    }
+
+    final rootClassId = graph.objects[heapRootIndex].classId;
+    final path = <HeapClassName>[];
+
+    while (shortestRetainers[nextObjectId] != heapRootIndex) {
+      nextObjectId = shortestRetainers[nextObjectId];
+      final classId = graph.objects[nextObjectId].classId;
+      assert(classId != rootClassId);
+      final className =
+          HeapClassName.fromHeapSnapshotClass(graph.classes[classId]);
+      path.add(className);
+    }
+
+    return PathFromRoot.fromPath(path);
+  }
+
+  factory PathFromRoot.fromPath(List<HeapClassName> path) {
+    if (path.isEmpty) return empty;
+    final existingInstance = instances.lookup(PathFromRoot._(path));
+    if (existingInstance != null) return existingInstance;
+
+    final newInstance = PathFromRoot._(List.unmodifiable(path));
+
+    instances.add(newInstance);
+    assert(() {
+      usage.stored++;
+      assert(instances.length == usage.stored);
+      return true;
+    }());
+    return newInstance;
+  }
+
+  @visibleForTesting
+  static Set<PathFromRoot> get instances => _instances ??= <PathFromRoot>{};
+  static Set<PathFromRoot>? _instances;
+
+  @visibleForTesting
+  static DebugRetainingPathUsage usage = DebugRetainingPathUsage();
+
+  static PathFromRoot empty = const PathFromRoot._empty();
+
+  /// If false, the [classes] is always empty.
+  ///
+  /// Is used to evaluate performance of calculations.
+  static bool calculateSetOfClasses = true;
```

Comment:
is this supposed to be true? also, if this is a debug flag, add it to development_helpers.dart instead